### PR TITLE
Added browserslist output.target and output.environment

### LINKED ISSUE #3968: Document webpack change: improve target option
*A pull request by @sokra was merged and maintainers requested a documentation change.*

See pull request: https://github.com/webpack/webpack/pull/11454

---

**What kind of change does this PR introduce?**
refactoring, feature

**Did you add tests for your changes?**
yes + existing tests

**Does this PR introduce a breaking change?**
minimal -> terser-webpack-plugin uses `ecmaVersion`

**What needs to be documented once your changes are merged?**
* `target` option now allows an array of target
* `target` no longer allows to pass an plugin, but it allows `false` instead
* `target` allows more complex targets with versions now
  * web: Web browser.
  * webworker: Web Worker, SharedWorker or Service Worker.
  * `[async-]node[X[.Y]]`: Node.js in version X.Y. The 'async-' prefix will load chunks asynchronously via 'fs' and 'vm' instead of 'require()'. Examples: node14.5, async-node10.
  * `electron[X[.Y]]-main/preload/renderer`: Electron in version X.Y. Script is running in main, preload resp. renderer context.
  * `nwjs[X[.Y]]` / `node-webkit[X[.Y]]`: NW.js in version X.Y.
  * `esX`: EcmaScript in this version. Examples: es5, es2020.
  * In future more targets will be added e. g. `chrome58` or `browserslist`
* The versions allow to infer features of the platform like if arrow functions are supported or not.
* When multiple targets are passed the common subset of features is used
* `output.ecmaVersion` has been replaced with `output.environment.*` and a list of boolean flags for features used by webpack (e. g. arrowFunction).
* When no information about the environment features is provided, we assume ES2015

### LINKED ISSUE #4009: Document webpack change: make browserslist the default target when browserslist config is available
*A pull request by @sokra was merged and maintainers requested a documentation change.*

See pull request: https://github.com/webpack/webpack/pull/11541

---

**What kind of change does this PR introduce?**
defaults change

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
yes

**What needs to be documented once your changes are merged?**
when browserslist info is available, this will be used by default

### PR DESCRIPTION
**Pull request by smelukov**
Added browserslist output.target and output.environment
closes #4009

* ~The versions allow to infer features of the platform like if arrow functions are supported or not.~
* ~`output.ecmaVersion` has been replaced with `output.environment.*` and a list of boolean flags for features used by webpack (e. g. arrowFunction).~

### CODE REVIEW

File: src/content/configuration/target.md [hunk 1]
```diff
@@ -10,13 +10,14 @@ contributors:
...
   - EugeneHlushko
+  - smelukov
 ---
 
 webpack can compile for multiple environments or _targets_. To understand what a `target` is in detail, read through [the targets concept page](/concepts/targets/).
 
 ## `target`
 
-`string` `function (compiler)`
+`string` `function (compiler)` = `"browserslist"` or `"web"` if no browserslist-config found
```
Comment (EugeneHlushko) [inline src/content/configuration/target.md:20]: complex, conditional default should in the description
```suggestion
`string` `function (compiler) => string`
```

File: src/content/configuration/target.md [hunk 2]
```diff
@@ -10,13 +10,14 @@ contributors:
...
 
 Instructs webpack to target a specific environment.
```
Comment (EugeneHlushko) [inline src/content/configuration/target.md:22]:
```suggestion
Instructs webpack to target a specific environment. Defaults to `'browserslist'` or to `'web'` when no browserslist-config was found.
```

**Review by EugeneHlushko — CHANGES_REQUESTED** [review]

**Comment by sokra** [comment]
@smelukov hey cool to see to working on that. Do you want to merge the other `target` changes into the PR?

see https://github.com/webpack/webpack.js.org/issues/3968

**Comment by smelukov** [comment]
@sokra ok, I'll do it on this PR ;)

**Comment by sokra** [comment]
Cool thanks

File: src/content/configuration/target.md [hunk 3]
```diff
@@ -35,9 +36,24 @@ Option                | Description
...
 `webworker`           | Compile as WebWorker
+`browserslist`        | Infer a platform and the ES-features from a browserslist-config
 
 For example, when the `target` is set to `"electron-main"`, webpack includes multiple electron specific variables. For more information on which templates and externals are used, you can refer to webpack's [source code](https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsApply.js#L148-L183).
 
+#### `browserslist`
+
+If a project has a browserslist config, then webpack will use it for:
+
+- determinate ES-features that may be used to generate a runtime-code (all the chunks and modules are wrapped by runtime code).
```
Comment (montogeek) [inline src/content/configuration/target.md:47]:
```suggestion
- Determinate ES-features that may be used to generate a runtime-code (all the chunks and modules are wrapped by runtime code).
```

File: src/content/configuration/target.md [hunk 4]
```diff
@@ -35,9 +36,24 @@ Option                | Description
...
+- infer an environment (e.g: `last 2 node versions` the same as `target: "node"` with some [`output.environment`](/configuration/output/#outputenvironment) settings)
```
Comment (montogeek) [inline src/content/configuration/target.md:48]:
```suggestion
- Infer an environment (e.g: `last 2 node versions` the same as `target: "node"` with some [`output.environment`](/configuration/output/#outputenvironment) settings)
```

Comment (montogeek) [inline src/content/configuration/target.md:48]:
```suggestion
- Infer an environment (e.g: `last 2 node versions` the same as `target: "node"` with some [`output.environment`](/configuration/output/#outputenvironment) settings).
```

File: src/content/configuration/target.md [hunk 5]
```diff
@@ -10,15 +10,16 @@ contributors:
...
 
 webpack can compile for multiple environments or _targets_. To understand what a `target` is in detail, read through [the targets concept page](/concepts/targets/).
 
 ## `target`
 
-`string` `function (compiler)`
+`string` `function (compiler) => string`
 
-Instructs webpack to target a specific environment.
+Instructs webpack to target a specific environment. Defaults to `'browserslist'` or to `'web'` when no browserslist-config was found.
```
Comment (montogeek) [inline src/content/configuration/target.md:22]:
```suggestion
Instructs webpack to target a specific environment. Defaults to `'browserslist'` or to `'web'` when no browserslist configuration was found.
```

File: src/content/configuration/output.md [hunk 6]
```diff
@@ -1200,25 +1200,33 @@ module.exports = {
...
+      bigIntLiteral: false,
+      // The environment supports const and let for variable declarations.
+      const: true,
+      // The environment supports destructuring ('{ a, b } = obj').
+      destructuring: true,
+      // The environment supports an async import() function to import EcmaScript modules.
+      dynamicImport: false,
+      // The environment supports 'for of' iteration ('for (const x of array) { ... }').
+      forOf: true,
+      // The environment supports EcmaScript Module syntax to import EcmaScript modules (import ... from '...').
```
Comment (montogeek) [inline src/content/configuration/output.md:1223]:
```suggestion
      // The environment supports ECMAScript Module syntax to import ECMAScript modules (import ... from '...').
```

**Review by montogeek — APPROVED** [review]

**Review by sokra — APPROVED** [review]
There are still open things for `target`, but the changes are fine. More improvements in separate PR.

**Comment by sokra** [comment]
Thanks

**Pull request merged**